### PR TITLE
added SoftNearestNeighborRegularizedCrossEntropy from 2019

### PR DESCRIPTION
**Pull request by nickfrosst**
added loss from Analyzing and Improving Representations with the Soft Nearest Neighbor Loss

### CODE REVIEW

**Comment by npapernot** [comment]
It looks like some of the lines are above 120 characters, so this will likely fail format tests: see https://github.com/tensorflow/cleverhans#contributing

File: cleverhans/loss.py [hunk 1]
```diff
@@ -329,3 +334,100 @@ def fprop(self, x, y, **kwargs):
                   "MixUp. LossFeaturePairing may be removed "
                   "on or after 2019-03-06.")
     return loss
+
+
+class SNNLCrossEntropy(CrossEntropy):
```
Comment (goodfeli) [inline cleverhans/loss.py]: Please add a docstring.
A few things that should go in the docstring:
  - Specification that this stands for Soft Nearest Neighbor + whatever the L stands for
  - Link to the paper

File: cleverhans/loss.py [hunk 2]
```diff
@@ -329,3 +334,100 @@ def fprop(self, x, y, **kwargs):
...
+  EPSILON = 0.00001
```
Comment (goodfeli) [inline cleverhans/loss.py]: Is it possible to use a different name, or a more descriptive name? Throughout most of the library we use `eps` to mean the size of the norm ball for a threat model.

File: cleverhans/loss.py (hunk 1 above)
Comment (goodfeli) [reply cleverhans/loss.py]: The docstring should also include ":param" declarations of the arguments of the `__init__` method

File: cleverhans/loss.py [hunk 3]
```diff
@@ -329,3 +334,100 @@ def fprop(self, x, y, **kwargs):
...
+    self.factor = factor
+    self.optimize_temperature = optimize_temperature
+    self.cos_distance = cos_distance
+    self.layer_names = layer_names
+    if not layer_names:
+      # omit the final classification layer
+      self.layer_names = model.get_layer_names()[:-1]
+
+  @staticmethod
+  def pairwise_euclid_distance(A, B):
```
Comment (goodfeli) [inline cleverhans/loss.py]: please add docstring

File: cleverhans/loss.py [hunk 4]
```diff
@@ -329,3 +334,100 @@ def fprop(self, x, y, **kwargs):
...
+    sqr_norm_A = tf.reshape(tf.reduce_sum(tf.pow(A, 2), 1), [1, batchA])
+    sqr_norm_B = tf.reshape(tf.reduce_sum(tf.pow(B, 2), 1), [batchB, 1])
+    inner_prod = tf.matmul(B, A, transpose_b=True)
+
+    tile_1 = tf.tile(sqr_norm_A, [batchB, 1])
+    tile_2 = tf.tile(sqr_norm_B, [1, batchA])
+    return (tile_1 + tile_2 - 2 * inner_prod)
+
+  @staticmethod
+  def pairwise_cos_distance(A, B):
```
Comment (goodfeli) [inline cleverhans/loss.py]: please add docstring

File: cleverhans/loss.py [hunk 5]
```diff
@@ -329,3 +334,100 @@ def fprop(self, x, y, **kwargs):
...
+    normalized_A = tf.nn.l2_normalize(A, dim=1)
+    normalized_B = tf.nn.l2_normalize(B, dim=1)
+    prod = tf.matmul(normalized_A, normalized_B, adjoint_b=True)
+    return 1 - prod
+
+  @staticmethod
+  def fits(A, B, temp, cos_distance):
```
Comment (goodfeli) [inline cleverhans/loss.py]: please add docstring

File: cleverhans/loss.py [hunk 6]
```diff
@@ -329,3 +334,100 @@ def fprop(self, x, y, **kwargs):
...
+    if cos_distance:
+      distance_matrix = SNNLCrossEntropy.pairwise_cos_distance(A, B)
+    else:
+      distance_matrix = SNNLCrossEntropy.pairwise_euclid_distance(A, B)
+    return tf.exp(-(distance_matrix / temp))
+
+  @staticmethod
+  def pick_probability(x, temp, cos_distance):
```
Comment (goodfeli) [inline cleverhans/loss.py]: I'll quit writing "please add docstring", but please add docstrings to all the new methods / functions in the PR

File: cleverhans/loss.py [hunk 7]
```diff
@@ -329,3 +334,100 @@ def fprop(self, x, y, **kwargs):
...
+
+  @staticmethod
+  def SNNL(x, y, temp, cos_distance):
+    summed_masked_pick_prob = tf.reduce_sum(
+        SNNLCrossEntropy.masked_pick_probability(x, y, temp, cos_distance), 1)
+    return tf.reduce_mean(
+        -tf.log(SNNLCrossEntropy.EPSILON + summed_masked_pick_prob))
+
+  @staticmethod
+  def optimizied_temp_SNNL(x, y, initial_temp, cos_distance):
```
Comment (goodfeli) [inline cleverhans/loss.py]: typo "optimizied"

**Comment by nickfrosst** [comment]
Thanks for reviewing! Also yeah I can add a model that uses it. I wrote one
already to test it locally so it should be easy :) I'll add the fixes in
soon.

**Comment by nickfrosst** [comment]
here are two examples of the output of the model i added to the model zoo. it just demonstrates the class clusters in adversarial gradient. 

adversarial_gradients_snnl_factor_-10 0
adversarial_gradients_snnl_factor_0 0

**Comment by npapernot** [comment]
OK to ignore the CLA bot
